File: src/main/java/ru/mrgrd56/api/image/ImageGenerationService.java
package ru.mrgrd56.api.image;

import org.springframework.stereotype.Service;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

@Service
public class ImageGenerationService {
    public byte[] generateFilledImage(
            Color color,
            Dimension size,
            Integer borderRadius
    ) {
        validateSize(size);
        validateBorderRadius(borderRadius);

        int width = size.width;
        int height = size.height;

        var image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D graphics = image.createGraphics();
        graphics.setColor(color);
        if (borderRadius == null) {
            graphics.fillRect(0, 0, width, height);
        } else {
            graphics.fillRoundRect(0, 0, width, height, borderRadius, borderRadius);
        }

        var result = new ByteArrayOutputStream();
        try {
            ImageIO.write(image, "png", result);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return result.toByteArray();
    }

    public Color parseColor(String input) throws IllegalArgumentException {
        var result = Color.getColor(input);
        if (result == null) {
            throw new IllegalArgumentException("Invalid color provided");
        }

        return result;
    }

    public Dimension parseSize(String input) throws IllegalArgumentException {
        String[] parts = input.split("[xх]", 2);
        if (parts.length == 1) {
            int size = Integer.parseInt(parts[0]);
            return new Dimension(size, size);
        }
        if (parts.length == 2) {
            int width = Integer.parseInt(parts[0]);
            int height = Integer.parseInt(parts[1]);
            return new Dimension(width, height);
        }

        throw new IllegalArgumentException("Invalid size (s) provided");
    }

    private void validateSize(Dimension dimension) throws IllegalArgumentException {
        if (dimension.width <= 0 || dimension.width > 3000 || dimension.height < 0 || dimension.height > 3000) {
            throw new IllegalArgumentException("Invalid size (s) provided: both sides must be in range [1; 3000]");
        }
    }

    private void validateBorderRadius(Integer borderRadius) {
        if (borderRadius == null) {
            return;
        }

        if (borderRadius < 0) {
            throw new IllegalArgumentException("Invalid borderRadius (r) provided: the value must not be less than 0");
        }
    }
}

File: src/main/kotlin/ru/mrgrd56/api/image/ImageGenerationService.kt
package ru.mrgrd56.api.image

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import org.springframework.stereotype.Service
import java.awt.Color
import java.awt.Dimension
import java.awt.image.BufferedImage
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.concurrent.TimeUnit
import javax.imageio.ImageIO

@Service
class ImageGenerationService {
    private val filledImagesCache: Cache<String, ByteArray> =
        Caffeine.newBuilder()
            .maximumSize(512)
            .expireAfterAccess(1, TimeUnit.HOURS)
            .build()

    fun generateFilledImage(
        color: Color,
        size: Dimension,
        borderRadius: Int?
    ): ByteArray {
        val hash = "$color/$size/$borderRadius"

        filledImagesCache.getIfPresent(hash)?.let {
            return it
        }

        validateSize(size)
        validateBorderRadius(borderRadius)

        val width = size.width
        val height = size.height
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)

        val graphics = image.createGraphics()
        graphics.color = color

        if (borderRadius == null) {
            graphics.fillRect(0, 0, width, height)
        } else {
            graphics.fillRoundRect(0, 0, width, height, borderRadius, borderRadius)
        }

        ByteArrayOutputStream().use { output ->
            try {
                ImageIO.write(image, "png", output)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

            val result = output.toByteArray()

            if (width <= 512 && height <= 512) {
                filledImagesCache.put(hash, result)
            }

            return result
        }
    }

    @Throws(IllegalArgumentException::class)
    fun parseColor(input: String?): Color {
        return Color.decode(input) ?: throw IllegalArgumentException("Invalid color provided")
    }

    @Throws(IllegalArgumentException::class)
    fun parseSize(input: String): Dimension {
        val parts = input.split("[xх]".toRegex(), limit = 2)

        return when (parts.size) {
            1 -> {
                val size = parts[0].toInt()
                Dimension(size, size)
            }

            2 -> {
                val width = parts[0].toInt()
                val height = parts[1].toInt()
                Dimension(width, height)
            }

            else -> throw IllegalArgumentException("Invalid size (s) provided")
        }
    }

    @Throws(IllegalArgumentException::class)
    private fun validateSize(dimension: Dimension) {
        require(dimension.width in 1..3000 && dimension.height in 0..3000) {
            "Invalid size (s) provided: both sides must be in range [1; 3000]"
        }
    }

    @Throws(IllegalArgumentException::class)
    private fun validateBorderRadius(borderRadius: Int?) {
        borderRadius?.let {
            require(it >= 0) { "Invalid borderRadius (r) provided: the value must not be less than 0" }
        }
    }
}
